os/lib/cbor.c

```diff
@@ -0,0 +1,434 @@
+/*
+ * Copyright (c) 2018, SICS, RISE AB
+ * Copyright (c) 2023, Uppsala universitet
+ * Copyright (c) 2024, Siemens AG
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the Institute nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+/**
+ * \addtogroup cbor
+ * @{
+ * \file
+ *         CBOR implementation.
+ */
+
+#include "lib/cbor.h"
+#include <string.h>
+
+/*---------------------------------------------------------------------------*/
+void
+cbor_init_writer(cbor_writer_state_t *state,
+                 uint8_t *buffer, size_t buffer_size)
+{
+  state->buffer_head = buffer;
+  state->buffer = buffer;
+  state->buffer_size = buffer_size;
+  state->nesting_depth = CBOR_MAX_NESTING;
+}
+/*---------------------------------------------------------------------------*/
+size_t
+cbor_end_writer(cbor_writer_state_t *state)
+{
+  return (state->nesting_depth == CBOR_MAX_NESTING) && state->buffer
+         ? (size_t)(state->buffer - state->buffer_head)
+         : 0;
+}
+/*---------------------------------------------------------------------------*/
+void
+cbor_break_writer(cbor_writer_state_t *state)
+{
+  state->buffer = NULL;
+  state->buffer_size = 0;
+}
+/*---------------------------------------------------------------------------*/
+const uint8_t *
+cbor_peek_output(cbor_writer_state_t *state)
```

Comment:
I understand what this function does but I am unsure of what it can be used for. Can you give an example of when someone would like to use this function?

Reply:
I had to compute a hash over a map while writing. For the time being, I removed this function.